pkg/protocol/engine/attestation/slotattestation/manager.go

```diff
@@ -0,0 +1,249 @@
+	}
+	m.TriggerConstructed()
+
+	return m
+}
+
+func (m *Manager) Shutdown() {
+	if err := m.writeToDisk(); err != nil {
+		panic(err)
+	}
+	m.TriggerStopped()
+}
+
+// Get returns the attestations that are included in the commitment of the given slot.
+// If attestationOffset=3 and commitment is 10, then the returned attestations are blocks from 7 to 10 that commit to at least 7.
+func (m *Manager) Get(index iotago.SlotIndex) (*ads.Map[iotago.AccountID, iotago.Attestation, *iotago.AccountID, *iotago.Attestation], error) {
+	m.commitmentMutex.RLock()
+	defer m.commitmentMutex.RUnlock()
+
+	if index > m.lastCommittedSlot {
+		return nil, errors.Errorf("slot %d is newer than last committed slot %d", index, m.lastCommittedSlot)
+	}
+	cutoffIndex, isValid := m.computeAttestationCommitmentOffset(index)
+	if !isValid {
+		return nil, errors.Errorf("slot %d is smaller than attestation cutoffIndex %d thus we don't have attestations", index, cutoffIndex)
+	}
+
+	return m.adsMapStorage(cutoffIndex)
+}
+
+// AddAttestationFromBlock adds an attestation from a block to the future attestations (beyond the attestation window).
+func (m *Manager) AddAttestationFromBlock(block *blocks.Block) {
+	// Only track attestations of active committee members.
+	if _, exists := m.committeeFunc().Get(block.Block().IssuerID); !exists {
+		return
+	}
+
+	m.commitmentMutex.RLock()
+	defer m.commitmentMutex.RUnlock()
+
+	// We only care about attestations that are newer than the last committed slot.
+	if block.ID().Index() <= m.lastCommittedSlot {
+		return
+	}
+
+	newAttestation := iotago.NewAttestation(block.Block())
+
+	// We keep only the latest attestation for each committee member.
+	m.futureAttestations.Get(block.ID().Index(), true).Compute(block.Block().IssuerID, func(currentValue *iotago.Attestation, exists bool) *iotago.Attestation {
+		if !exists {
+			return newAttestation
+		}
+
+		// Replace the attestation only if the new one is greater.
+		if newAttestation.Compare(currentValue) == 1 {
+			return newAttestation
+		}
+
+		return currentValue
+	})
+}
+
+func (m *Manager) applyToPendingAttestations(attestation *iotago.Attestation, cutoffIndex iotago.SlotIndex) {
+	if attestation.SlotCommitmentID.Index() < cutoffIndex {
+		return
+	}
+
+	updated := true
+
+	updatedAttestation := m.pendingAttestations.Get(cutoffIndex, true).Compute(attestation.IssuerID, func(currentValue *iotago.Attestation, exists bool) *iotago.Attestation {
+		if !exists {
+			return attestation
+		}
+
+		// Replace the stored attestation only if the new one is greater.
+		if attestation.Compare(currentValue) == 1 {
+			return attestation
+		}
+
+		updated = false
+
+		return currentValue
+	})
+
+	// The new attestation is smaller or equal the previousAttestation. There's no need to update pendingAttestations.
+	if !updated {
+		return
+	}
+
+	for i := cutoffIndex; i <= updatedAttestation.SlotCommitmentID.Index(); i++ {
```

Comment:
This `for` basically makes sure that we add an Attestation not only to the committing slot, but all pending slots before that in case of a gap, right?

Reply:
This ensures that the "latest" attestation for each issuer is "applied" to all indeces before the one it commits to. If the attestation, say, committed to 5, it should be a valid attestation for all the slots before it.
If the `Compare` above doesn't take into account the commitment of the attestation itself, but only its issuing time, wouldn't you override a possibly later slot attestation here?

Reply:
You could have the same issuer having an attestation committing to 5 on slots 0-4 and an attestation committing to 3 on slot 5, provided that this latest attestation has a newer issuingTime.

Reply:
> but all pending slots before that in case of a gap
Exactly. A "vote" on slot 5 indirectly "votes" for all slots `<= 5`.

> If the Compare above doesn't take into account the commitment of the attestation itself
The `compare` takes the commitment index into account.